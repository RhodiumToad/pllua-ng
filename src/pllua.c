/*
 * pllua.c: PL/Lua NG call handler
 * By Andrew "RhodiumToad" Gierth, rhodiumtoad at postgresql.org
 * Based in some part on pllua by Luis Carvalho and others
 * License: MIT license or PostgreSQL licence
 */

#include "pllua.h"

#include "commands/event_trigger.h"
#include "commands/trigger.h"

PG_MODULE_MAGIC;

/*
 * Exposed interface
 *
 * see also _PG_init in init.c
 */
PGDLLEXPORT Datum pllua_validator(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum pllua_call_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum pllua_inline_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_validator(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_call_handler(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum plluau_inline_handler(PG_FUNCTION_ARGS);

static Datum pllua_common_call(FunctionCallInfo fcinfo, bool trusted);
static Datum pllua_common_inline(FunctionCallInfo fcinfo, bool trusted);
static Datum pllua_common_validator(FunctionCallInfo fcinfo, bool trusted);


/* Trusted entry points */

PG_FUNCTION_INFO_V1(pllua_validator);
Datum pllua_validator(PG_FUNCTION_ARGS)
{
	return pllua_common_validator(fcinfo, true);
}

PG_FUNCTION_INFO_V1(pllua_call_handler);
Datum pllua_call_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_call(fcinfo, true);
}

PG_FUNCTION_INFO_V1(pllua_inline_handler);
Datum pllua_inline_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_inline(fcinfo, true);
}

/* Untrusted entry points */

PG_FUNCTION_INFO_V1(plluau_validator);
Datum plluau_validator(PG_FUNCTION_ARGS)
{
	return pllua_common_validator(fcinfo, false);
}

PG_FUNCTION_INFO_V1(plluau_call_handler);
Datum plluau_call_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_call(fcinfo, false);
}

PG_FUNCTION_INFO_V1(plluau_inline_handler);
Datum plluau_inline_handler(PG_FUNCTION_ARGS)
{
	return pllua_common_inline(fcinfo, false);
}

/* Common implementations */

static void pllua_entry_stack_check(void)
{
	check_stack_depth();
}

Datum pllua_common_call(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *volatile interp = NULL;
	pllua_activation_record act;
	pllua_func_activation *funcact = (fcinfo->flinfo) ? fcinfo->flinfo->fn_extra : NULL;
	ErrorContextCallback ecxt;

	pllua_entry_stack_check();

	act.fcinfo = fcinfo;
	act.retval = (Datum) 0;
	act.atomic = true;
	act.trusted = trusted;
	act.cblock = NULL;
	act.validate_func = InvalidOid;
	act.interp = NULL;
	act.active_error = LUA_REFNIL;
	act.err_text = NULL;

#if PG_VERSION_NUM >= 110000
	if (fcinfo->context && IsA(fcinfo->context, CallContext))
		act.atomic = castNode(CallContext, fcinfo->context)->atomic;
#endif

	pllua_setcontext(NULL, PLLUA_CONTEXT_PG);

	/*
	 * this catch block exists to save/restore the error context stack and
	 * allow cleanup of our internal error state when returning to PG proper
	 */
	PG_TRY();
	{
		ecxt.callback = pllua_error_callback;
		ecxt.arg = &act;
		ecxt.previous = error_context_stack;
		error_context_stack = &ecxt;

		if (funcact && funcact->thread)
			act.interp = funcact->interp;
		else
			act.interp = pllua_getstate(trusted, &act);

		interp = act.interp;

		if (funcact && funcact->thread)
		{
			/*
			 * We're resuming a value-per-call SRF, so we bypass almost
			 * everything since we don't want to, for example, compile a new
			 * version of the function halfway through a result set. We know
			 * we're in a non-first-row condition if there's an existing thread
			 * in the function activation.
			 */
			pllua_initial_protected_call(act.interp, pllua_resume_function, &act);
		}
		else if (CALLED_AS_TRIGGER(fcinfo))
			pllua_initial_protected_call(act.interp, pllua_call_trigger, &act);
		else if (CALLED_AS_EVENT_TRIGGER(fcinfo))
			pllua_initial_protected_call(act.interp, pllua_call_event_trigger, &act);
		else
			pllua_initial_protected_call(act.interp, pllua_call_function, &act);
	}
	PG_CATCH();
	{
		if (interp)
			pllua_error_cleanup(interp, &act);
		PG_RE_THROW();
	}
	PG_END_TRY();

	return act.retval;
}

Datum pllua_common_validator(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *volatile interp = NULL;
	pllua_activation_record act;
	Oid funcoid = PG_GETARG_OID(0);
	ErrorContextCallback ecxt;

	/* security checks */
	if (!CheckFunctionValidatorAccess(fcinfo->flinfo->fn_oid, funcoid))
		PG_RETURN_VOID();

	/*
	 * This doesn't need a stack check because the validator is careful
	 * not to execute user-supplied Lua code.
	 */

	act.fcinfo = NULL;
	act.retval = (Datum) 0;
	act.atomic = true;
	act.trusted = trusted;
	act.cblock = NULL;
	act.validate_func = funcoid;
	act.interp = NULL;
	act.active_error = LUA_REFNIL;
	act.err_text = NULL;

	pllua_setcontext(NULL, PLLUA_CONTEXT_PG);

	/*
	 * this catch block exists to save/restore the error context stack and
	 * allow cleanup of our internal error state when returning to PG proper
	 */
	PG_TRY();
	{
		ecxt.callback = pllua_error_callback;
		ecxt.arg = &act;
		ecxt.previous = error_context_stack;
		error_context_stack = &ecxt;

		interp = act.interp = pllua_getstate(trusted, &act);

		pllua_initial_protected_call(act.interp, pllua_validate, &act);
	}
	PG_CATCH();
	{
		if (interp)
			pllua_error_cleanup(interp, &act);
		PG_RE_THROW();
	}
	PG_END_TRY();

	PG_RETURN_VOID();
}

Datum pllua_common_inline(FunctionCallInfo fcinfo, bool trusted)
{
	pllua_interpreter *volatile interp = NULL;
	pllua_activation_record act;
	ErrorContextCallback ecxt;

	pllua_entry_stack_check();

	act.fcinfo = NULL;
	act.retval = (Datum) 0;
	act.atomic = true;
	act.trusted = trusted;
	act.cblock = (InlineCodeBlock *) PG_GETARG_POINTER(0);
	act.validate_func = InvalidOid;
	act.interp = NULL;
	act.active_error = LUA_REFNIL;
	act.err_text = "inline block entry";

#if PG_VERSION_NUM >= 110000
	act.atomic = act.cblock->atomic;
#endif

	pllua_setcontext(NULL, PLLUA_CONTEXT_PG);

	/* probably excess paranoia */
	if (act.cblock->langIsTrusted != act.trusted)
		elog(ERROR, "trusted state mismatch");

	/*
	 * this catch block exists to save/restore the error context stack and
	 * allow cleanup of our internal error state when returning to PG proper
	 */
	PG_TRY();
	{
		ecxt.callback = pllua_error_callback;
		ecxt.arg = &act;
		ecxt.previous = error_context_stack;
		error_context_stack = &ecxt;

		interp = act.interp = pllua_getstate(trusted, &act);

		pllua_initial_protected_call(act.interp, pllua_call_inline, &act);
	}
	PG_CATCH();
	{
		if (interp)
			pllua_error_cleanup(interp, &act);
		PG_RE_THROW();
	}
	PG_END_TRY();

	PG_RETURN_VOID();
}
