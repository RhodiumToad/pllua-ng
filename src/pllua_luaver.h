/* pllua_luaver.h */

#ifndef PLLUA_LUAVER_H
#define PLLUA_LUAVER_H

#ifndef LUAJIT_VERSION_NUM
#define LUAJIT_VERSION_NUM 0
#endif

/*
 * We can only use 5.1 environments in a compatible way to 5.2+ uservalues if
 * we forcibly set an environment on every userdata we create.
 */
#if LUA_VERSION_NUM == 501
#define MANDATORY_USERVALUE 1
#else
#define MANDATORY_USERVALUE 0
#endif

/*
 * pllua_pushcfunction must absolutely not throw error.
 *
 * In Lua 5.4+ (and also 5.3.5+, but we don't check that) it's safe to just use
 * lua_pushcfunction, but in 5.1 (by design) and 5.3.[34] (due to bugs) we must
 * instead arrange to store all the function values in the registry, and use a
 * rawget to fetch them.
 */
#if LUA_VERSION_NUM > 503

#define pllua_pushcfunction(L_,f_) lua_pushcfunction(L_,f_)

#else

#define pllua_pushcfunction(L_,f_) do {					\
		int rc_ PG_USED_FOR_ASSERTS_ONLY;				\
		rc_ = lua_rawgetp((L_),LUA_REGISTRYINDEX,(f_));	\
		Assert(rc_==LUA_TFUNCTION); } while(0)

#endif

/*
 * used to label functions that need registration despite not being
 * directly passed to pcall or cpcall; the first arg is unused
 */
#define pllua_register_cfunc(L_, f_) (f_)

/*
 * Function to use to set an environment on a code chunk.
 */
#if LUA_VERSION_NUM == 501
#define pllua_set_environment(L_,i_) lua_setfenv(L, i_)
#else
#define pllua_set_environment(L_,i_) lua_setupvalue(L_, i_, 1)
#endif

/*
 * Handle API differences for lua_resume by emulating the 5.4 API on earlier
 * versions. Also fake out the warning system on earlier versions, and provide
 * a minimal emulation of <close> that works for normal exits (leaving error
 * exits to be cleaned up by the GC, but that can't be helped).
 */
#if LUA_VERSION_NUM < 504

static inline int
pllua_resume(lua_State *L, lua_State *from, int nargs, int *nret)
{
#if LUA_VERSION_NUM == 501
	int rc = (lua_resume)(L, nargs);
#else
	int rc = (lua_resume)(L, from, nargs);
#endif
	*nret = lua_gettop(L);
	return rc;
}
#define lua_resume(L_,f_,a_,r_) (pllua_resume(L_,f_,a_,r_))

#define lua_setwarnf(L_, f_, p_) ((void)(f_))

#define lua_setcstacklimit(L_, n_) (200)

#define lua_resetthread(L_) (LUA_OK)

#define PLLUA_WARNBUF_SIZE 4

#define lua_toclose(L_, i_) ((void)0)

static inline void
pllua_closevar(lua_State *L, int idx)
{
	if (lua_toboolean(L, idx)
		&& luaL_callmeta(L, idx, "__close"))
		lua_pop(L, 1);
}

#else

#define PLLUA_WARNBUF_SIZE 1000

#define pllua_closevar(L_, i_) ((void)0)

#endif /* LUA_VERSION_NUM < 504 */

#endif
