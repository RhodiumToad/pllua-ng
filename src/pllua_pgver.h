/*
 * pllua_pgver.h
 */

#ifndef PLLUA_PGVER_H
#define PLLUA_PGVER_H

/* PG version cruft */

/*
 * It sucks to depend on point release version, but some values we want to use
 * in #if directives are not compile-time constant in older point releases.
 * So for unfixed point releases, forcibly fix things here.
 *
 * Fortunately this matters only at compile time.
 */
#if (PG_VERSION_NUM >= 90500 && PG_VERSION_NUM < 90510) \
	|| (PG_VERSION_NUM >= 90600 && PG_VERSION_NUM < 90606) \
	|| (PG_VERSION_NUM >= 100000 && PG_VERSION_NUM < 100001)

#undef INT64CONST
#undef UINT64CONST
#if defined(HAVE_LONG_INT_64)
#define INT64CONST(x)  (x##L)
#define UINT64CONST(x) (x##UL)
#elif defined(HAVE_LONG_LONG_INT_64)
#define INT64CONST(x)  (x##LL)
#define UINT64CONST(x) (x##ULL)
#else
#error must have a working 64-bit integer datatype
#endif

#ifdef PG_INT64_MIN
#undef PG_INT64_MIN
#endif
#ifdef PG_INT64_MAX
#undef PG_INT64_MAX
#endif

#endif /* CONST_MAX_HACK */

#ifndef PG_INT64_MIN
#define PG_INT64_MIN	(-INT64CONST(0x7FFFFFFFFFFFFFFF) - 1)
#endif
#ifndef PG_INT64_MAX
#define PG_INT64_MAX	INT64CONST(0x7FFFFFFFFFFFFFFF)
#endif

/* float4 is always by value in pg >= 13. */
#if PG_VERSION_NUM >= 130000
#ifndef USE_FLOAT4_BYVAL
#define USE_FLOAT4_BYVAL 1
#endif
#endif

/*
 * CommandTag in pg13+ is an enum, not a string. GetCommandTagName returns the
 * old name, but obviously didn't exist in previous versions.
 */
#if PG_VERSION_NUM < 130000
#define GetCommandTagName(t_) (t_)
#endif

/* RIP, oids. */
#if PG_VERSION_NUM >= 120000
#define TupleDescHasOids(tupdesc) (false)
#define IsObjectIdAttributeNumber(a) (false)
/*
 * since hasoid is always supposed to be false thanks to the above, any
 * references to the below macros should be unreachable
 */
#define HeapTupleHeaderGetOid(h) (AssertMacro(false), InvalidOid)
#define HeapTupleHeaderSetOid(h,o) do { (void) (h); (void) (o); Assert(false); } while (0)
#define HeapTupleSetOid(h,o) do { (void) (h); (void) (o); Assert(false); } while (0)
#else
#define TupleDescHasOids(tupdesc) ((tupdesc)->tdhasoid)
#define IsObjectIdAttributeNumber(a) ((a) == ObjectIdAttributeNumber)
#endif

/* cope with variable-length fcinfo in pg12 */
#if PG_VERSION_NUM < 120000
#define LOCAL_FCINFO(name_,nargs_) \
	FunctionCallInfoData name_##data; \
	FunctionCallInfo name_ = &name_##data

#define LFCI_ARG_VALUE(fci_,n_) ((fci_)->arg[n_])
#define LFCI_ARGISNULL(fci_,n_) ((fci_)->argnull[n_])
#else
#define LFCI_ARG_VALUE(fci_,n_) ((fci_)->args[n_].value)
#define LFCI_ARGISNULL(fci_,n_) ((fci_)->args[n_].isnull)
#endif

/* TupleDesc structure change */
#if PG_VERSION_NUM < 100000
#define TupleDescAttr(tupdesc, i) ((tupdesc)->attrs[(i)])
#endif

/* AllocSetContextCreate API changes */
#if PG_VERSION_NUM < 110000
#define AllocSetContextCreateInternal AllocSetContextCreate
#elif PG_VERSION_NUM < 120000
#define AllocSetContextCreateInternal AllocSetContextCreateExtended
#endif
/*
 * Protect against backpatching or lack thereof.
 */
#ifndef ALLOCSET_DEFAULT_SIZES
#define ALLOCSET_DEFAULT_SIZES \
	ALLOCSET_DEFAULT_MINSIZE, ALLOCSET_DEFAULT_INITSIZE, ALLOCSET_DEFAULT_MAXSIZE
#endif
#ifndef ALLOCSET_SMALL_SIZES
#define ALLOCSET_SMALL_SIZES \
	ALLOCSET_SMALL_MINSIZE, ALLOCSET_SMALL_INITSIZE, ALLOCSET_SMALL_MAXSIZE
#endif
#ifndef ALLOCSET_START_SMALL_SIZES
#define ALLOCSET_START_SMALL_SIZES \
	ALLOCSET_SMALL_MINSIZE, ALLOCSET_SMALL_INITSIZE, ALLOCSET_DEFAULT_MAXSIZE
#endif

/* We want a way to do noinline, but old PGs don't have it. */

#ifndef __has_builtin
#define __has_builtin(x_) 0
#endif
#ifndef __has_attribute
#define __has_attribute(x_) 0
#endif

#if defined(pg_noinline)
#define pllua_noinline pg_noinline
#elif __has_attribute(noinline) || (defined(__GNUC__) && __GNUC__ > 2) || defined(__SUNPRO_C) || defined(__IBMC__)
#define pllua_noinline __attribute__((noinline))
/* msvc via declspec */
#elif defined(_MSC_VER)
#define pllua_noinline __declspec(noinline)
#else
#define pllua_noinline
#endif

/* and likewise for unlikely() */
#if !defined(unlikely)

#if !defined(__builtin_expect) && !defined(__GNUC__) && !__has_builtin(__builtin_expect)
#define __builtin_expect(x_,y_) (x_)
#endif
#define likely(x)	(__builtin_expect(!!(x), 1))
#define unlikely(x)	(__builtin_expect(!!(x), 0))

#endif /* unlikely */

#endif /* PLLUA_PGVER_H */
