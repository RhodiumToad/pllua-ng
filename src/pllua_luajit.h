/* pllua_luajit.h */

#ifndef PLLUA_LUAJIT_H
#define PLLUA_LUAJIT_H

#include <luajit.h>

/*
 * these should be the largest and smallest integers representable exactly in
 * a lua number (note, NOT necessarily a lua_Integer); must be compile-time
 * constant.
 */
#define PLLUA_MAX_INT_NUM (INT64CONST(9007199254740991))
#define PLLUA_MIN_INT_NUM (-INT64CONST(9007199254740991))

/*
 * these must be the exact bounds of a lua_Integer.
 */
#define LUA_MAXINTEGER PTRDIFF_MAX
#define LUA_MININTEGER PTRDIFF_MIN

#include "pllua_luaver.h"

/*
 * Parts of the lua 5.1 compatibility cruft here is derived from the
 * lua-compat-5.3 project, which is licensed under the same terms as this
 * project and carries the following copyright:
 *
 * Copyright (c) 2015 Kepler Project.
 */
extern const char *luaL_tolstring(lua_State *L, int nd, size_t *len);

static inline int lua_absindex(lua_State *L, int nd)
{
	return (nd < 0 && nd > LUA_REGISTRYINDEX) ? nd + lua_gettop(L) + 1 : nd;
}

static inline int lua_rawgetp(lua_State *L, int nd, void *p)
{
	int tnd = lua_absindex(L, nd);
	lua_pushlightuserdata(L, p);
	lua_rawget(L, tnd);
	return lua_type(L, -1);
}

static inline void lua_rawsetp(lua_State *L, int nd, void *p)
{
	int tnd = lua_absindex(L, nd);
	lua_pushlightuserdata(L, p);
	lua_insert(L, -2);
	lua_rawset(L, tnd);
}

static inline int lua_geti(lua_State *L, int nd, lua_Integer i)
{
	int tnd = lua_absindex(L, nd);
	lua_pushinteger(L, i);
	lua_gettable(L, tnd);
	return lua_type(L, -1);
}

static inline void lua_seti(lua_State *L, int nd, lua_Integer i)
{
	int tnd = lua_absindex(L, nd);
	lua_pushinteger(L, i);
	lua_insert(L, -2);
	lua_settable(L, tnd);
}

#define lua_rawgeti(L_,nd_,i_) ((lua_rawgeti)(L_,nd_,i_),lua_type(L_, -1))

#define lua_rawget(L_,nd_) ((lua_rawget)(L_,nd_),lua_type(L_, -1))

#define lua_getfield(L_,nd_,i_) ((lua_getfield)(L_,nd_,i_),lua_type(L_, -1))

#define lua_gettable(L_,nd_) ((lua_gettable)(L_,nd_),lua_type(L_, -1))

#define luaL_getmetafield(L_,nd_,f_) ((luaL_getmetafield)(L_,nd_,f_) ? lua_type(L_, -1) : LUA_TNIL)

/* luajit 2.1's version of this one is ok. */
#if LUAJIT_VERSION_NUM < 20100
static inline lua_Number lua_tonumberx(lua_State *L, int i, int *isnum)
{
	lua_Number n = lua_tonumber(L, i);
	if (isnum != NULL) {
		*isnum = (n != 0 || lua_isnumber(L, i));
	}
	return n;
}
#endif

/*
 * but for these we need to kill luajit's version and use ours: we depend on
 * isinteger/tointegerx/checkinteger accepting only actually integral values,
 * not rounded/truncated floats.
 */
#if LUAJIT_VERSION_NUM >= 20100
#define lua_isinteger pllua_isinteger
#define lua_tointegerx pllua_tointegerx
#endif

/* no luajit version of these is usable: */
#define luaL_checkinteger pllua_checkinteger
#define luaL_optinteger pllua_optinteger

static inline bool lua_isinteger(lua_State *L, int nd)
{
	if (lua_type(L, nd) == LUA_TNUMBER)
	{
		lua_Number n = lua_tonumber(L, nd);
		lua_Integer i = lua_tointeger(L, nd);
		if (i == n)
			return 1;
	}
	return 0;
}

static inline lua_Integer lua_tointegerx(lua_State *L, int i, int *isint)
{
	lua_Integer n = lua_tointeger(L, i);
	if (isint != NULL) {
		int isnum = 0;
		/* be careful, it might not be a number at all */
		if (n == lua_tonumberx(L, i, &isnum))
			*isint = isnum;
		else
			*isint = 0;
	}
	return n;
}

static inline lua_Integer luaL_checkinteger(lua_State *L, int i)
{
	int isint = 0;
	lua_Integer res = lua_tointegerx(L, i, &isint);
	if (!isint)
		luaL_argerror(L, i, "integer");
	return res;
}

static inline lua_Integer luaL_optinteger(lua_State *L, int i, lua_Integer def)
{
	if (lua_isnoneornil(L, i))
		return def;
	else
		return luaL_checkinteger(L, i);
}

/*
 * Miscellaneous functions
 */
#define lua_getuservalue(L_,nd_) (lua_getfenv(L_,nd_), lua_type(L_,-1))
#define lua_setuservalue(L_,nd_) lua_setfenv(L_,nd_)

#define lua_pushglobaltable(L_) lua_pushvalue((L_), LUA_GLOBALSINDEX)

#if LUAJIT_VERSION_NUM < 20100
#define luaL_setfuncs(L_,f_,u_) pllua_setfuncs(L_,f_,u_)
void pllua_setfuncs(lua_State *L, const luaL_Reg *reg, int nup);
#endif

#define luaL_getsubtable(L_,i_,n_) pllua_getsubtable(L_,i_,n_)
int pllua_getsubtable(lua_State *L, int i, const char *name);

#define luaL_requiref(L_,m_,f_,g_) pllua_requiref(L_,m_,f_,g_)
void pllua_requiref(lua_State *L, const char *modname, lua_CFunction openf, int glb);

#ifndef luaL_newlibtable
#define luaL_newlibtable(L, l)							\
  (lua_createtable((L), 0, sizeof((l))/sizeof(*(l))-1))
#define luaL_newlib(L, l) \
  (luaL_newlibtable((L), (l)), luaL_register((L), NULL, (l)))
#endif

#define LUA_OK 0

#if LUAJIT_VERSION_NUM > 0 && !defined(NO_LUAJIT)
#define LUA_TCDATA 10
#endif

#endif
