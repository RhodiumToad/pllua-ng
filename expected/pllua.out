--
CREATE EXTENSION pllua;
CREATE EXTENSION plluau;
\set VERBOSITY terse
-- smoke test
do language pllua $$ print "hello world!" $$;
INFO:  hello world!
do language plluau $$ print "hello world!" $$;
INFO:  hello world!
create function pg_temp.f1() returns text language pllua as $$ return "hello world" $$;
select pg_temp.f1();
     f1      
-------------
 hello world
(1 row)

create function pg_temp.f2() returns text language plluau as $$ return "hello world" $$;
select pg_temp.f2();
     f2      
-------------
 hello world
(1 row)

-- Rest of this file concentrates on simple tests of code paths in
-- compile, exec, and interpreter setup. Tests of other parts of the
-- module are separate.
-- validator
create function pg_temp."bad name"() returns text language pllua as $$ $$;
ERROR:  PL/Lua function name "bad name" is not a valid Lua identifier
create function pg_temp.f3("bad arg" integer) returns text language pllua as $$ $$;
ERROR:  PL/Lua argument name "bad arg" is not a valid Lua identifier
-- simple params and results (see types.sql for detailed checks)
create function pg_temp.f4(a integer) returns integer language pllua as $$ return a + 1 $$;
select pg_temp.f4(1);
 f4 
----
  2
(1 row)

create function pg_temp.f5(a text) returns text language pllua as $$ return a.."bar" $$;
select pg_temp.f5('foo');
   f5   
--------
 foobar
(1 row)

create function pg_temp.f6(a text, b integer) returns text language pllua as $$ return a..b $$;
select pg_temp.f6('foo',1);
  f6  
------
 foo1
(1 row)

-- try some polymorphism too
create function pg_temp.f7(a anyelement) returns anyelement language pllua as $$ return a $$;
select pg_temp.f7(text 'foo');
 f7  
-----
 foo
(1 row)

select pg_temp.f7(json '{"foo":1}');
    f7     
-----------
 {"foo":1}
(1 row)

--select pg_temp.f7(xml '<foo>bar</foo>'); -- don't bother with this, might be compiled out
select pg_temp.f7(varchar 'foo');
 f7  
-----
 foo
(1 row)

select 'x',pg_temp.f7('foo'::char(20)),'x';
 ?column? |          f7          | ?column? 
----------+----------------------+----------
 x        | foo                  | x
(1 row)

select pg_temp.f7(cstring 'foo');
 f7  
-----
 foo
(1 row)

select pg_temp.f7(name 'foo');
 f7  
-----
 foo
(1 row)

select pg_temp.f7(bytea 'foo\000bar');
        f7        
------------------
 \x666f6f00626172
(1 row)

select pg_temp.f7(smallint '2');
 f7 
----
  2
(1 row)

select pg_temp.f7(integer '2');
 f7 
----
  2
(1 row)

select pg_temp.f7(bigint '123456789012345');
       f7        
-----------------
 123456789012345
(1 row)

select pg_temp.f7(oid '10');
 f7 
----
 10
(1 row)

select pg_temp.f7(oid '4294967295');
     f7     
------------
 4294967295
(1 row)

select pg_temp.f7(true);
 f7 
----
 t
(1 row)

select pg_temp.f7(false);
 f7 
----
 f
(1 row)

select pg_temp.f7(1.5::float8);
 f7  
-----
 1.5
(1 row)

select pg_temp.f7(1.5::float4);
 f7  
-----
 1.5
(1 row)

-- variadics
create function pg_temp.f8(a text, variadic b integer[]) returns void language pllua as $$ print(a,type(b),b) $$;
select pg_temp.f8('foo', 1, 2, 3);
INFO:  foo	userdata	{1,2,3}
 f8 
----
 
(1 row)

create function pg_temp.f9(a integer, variadic b text[]) returns void language pllua as $$ print(a,type(b),b) $$;
select pg_temp.f9(1, 'foo', 'bar', 'baz');
INFO:  1	userdata	{foo,bar,baz}
 f9 
----
 
(1 row)

create function pg_temp.f10(a integer, variadic "any") returns void language pllua as $$ print(a,...) $$;
select pg_temp.f10(1, 'foo', 2, 'baz');
INFO:  1	foo	2	baz
 f10 
-----
 
(1 row)

-- SRF code paths
create function pg_temp.f11(a integer) returns setof text
  language pllua as $$ return $$;  -- 0 rows
select * from pg_temp.f11(1);
 f11 
-----
(0 rows)

create function pg_temp.f11b(a integer) returns setof text
  language pllua as $$ return 'foo' $$;  -- 1 row
select * from pg_temp.f11b(1);
 f11b 
------
 foo
(1 row)

create function pg_temp.f12(a integer) returns setof text
  language pllua as $$ coroutine.yield() $$;  -- 1 row, null
select * from pg_temp.f12(1);
 f12 
-----
 
(1 row)

create function pg_temp.f13(a integer) returns setof text
  language pllua as $$ for i = 1,a do coroutine.yield("row "..i) end $$;
select * from pg_temp.f13(4);
  f13  
-------
 row 1
 row 2
 row 3
 row 4
(4 rows)

create function pg_temp.f14(a integer, out x text, out y integer) returns setof record
  language pllua as $$ for i = 1,a do coroutine.yield("row "..i, i) end $$;
select * from pg_temp.f14(4);
   x   | y 
-------+---
 row 1 | 1
 row 2 | 2
 row 3 | 3
 row 4 | 4
(4 rows)

create function pg_temp.f15(a integer) returns table(x text, y integer)
  language pllua as $$ for i = 1,a do coroutine.yield("row "..i, i) end $$;
select * from pg_temp.f15(4);
   x   | y 
-------+---
 row 1 | 1
 row 2 | 2
 row 3 | 3
 row 4 | 4
(4 rows)

create function pg_temp.f16(a inout integer, x out text) returns setof record
  language pllua as $$ for i = 1,a do coroutine.yield(i, "row "..i) end $$;
select * from pg_temp.f16(4);
 a |   x   
---+-------
 1 | row 1
 2 | row 2
 3 | row 3
 4 | row 4
(4 rows)

-- SRF vs null returns
create function pg_temp.f16b(a integer) returns table(x text, y integer)
  language pllua as $$ coroutine.yield() $$;  -- 1 row, null
select * from pg_temp.f16b(1);
 x | y 
---+---
   |  
(1 row)

create function pg_temp.f16c(a integer) returns table(x text, y integer)
  language pllua as $$ coroutine.yield() for i = 1,a do coroutine.yield('foo',i) end $$;
select * from pg_temp.f16c(3);
  x  | y 
-----+---
     |  
 foo | 1
 foo | 2
 foo | 3
(4 rows)

-- compiler and validator code paths
do language pllua $$ _G.rdepth = 40 $$;  -- global var hack
-- This function will try and call itself at a point where it is visible
-- but has no definition interned yet; the recursive call will likewise
-- not see an interned definition and recurses again. without any limits
-- this would hit a stack depth check somewhere; we eat about 3 levels of
-- C function recursion inside lua each time, and that gets capped at 200.
-- We don't expect this to be actually useful, the test is just that we
-- don't crash.
create function pg_temp.f17(a integer) returns integer language pllua
  as $$
    return a
  end
  do
    if _G.rdepth > 0 then
      _G.rdepth = _G.rdepth - 1
      u = spi.execute("select pg_temp.f17(1)")
    end
$$;
select pg_temp.f17(1);
 f17 
-----
   1
(1 row)

create type pg_temp.t1 as (a integer, b text);
create function pg_temp.f18(a integer, b text) returns pg_temp.t1
  language pllua as $$ return a,b $$;
select * from pg_temp.f18(123,'foo');
  a  |  b  
-----+-----
 123 | foo
(1 row)

create function pg_temp.f19(a integer) returns text language pllua as $$ return 'foo '..a $$;
select pg_temp.f19(2);
  f19  
-------
 foo 2
(1 row)

create or replace function pg_temp.f19(a integer) returns text language pllua as $$ return 'bar '..a $$;
select pg_temp.f19(3);
  f19  
-------
 bar 3
(1 row)

-- trusted interpreter setup
-- check we really do have different interpreters
-- this is hard because we intentionally isolate trusted-language code
-- from the normal global env of its interpreter, so we would only be
-- able to verify isolation if we were able to break out of the
-- sandbox, which would rather defeat the point. We have to take the
-- outside view, by generating an interpreter-dependent value and
-- checking that it differs. The stringification of a closure, such as
-- server.error, suffices since this contains an interpreter-dependent
-- address (whereas base C functions do not differ between
-- interpreters in recent lua versions).
create function pg_temp.f20() returns text language pllua as $$ return tostring(spi.error) $$;
create function pg_temp.f21() returns text language plluau as $$ return tostring(spi.error) $$;
select pg_temp.f20() as a intersect select pg_temp.f21();  -- should be empty
 a 
---
(0 rows)

-- check the global table
do language pllua $$
  local gk = { "io", "dofile", "debug" }  -- must not exist
  for i = 1,#gk do print(gk[i],type(_G[gk[i]])) end
$$;
INFO:  io	nil
INFO:  dofile	nil
INFO:  debug	nil
do language plluau $$
  local gk = { "io", "dofile" }  -- probably exist
  for i = 1,#gk do print(gk[i],type(_G[gk[i]])) end
$$;
INFO:  io	table
INFO:  dofile	function
-- check that trusted gets only the restricted os module, even from
-- require
do language pllua $$
  local os = require 'os'
  local gk = { "time", "difftime", "execute", "getenv", "exit" }
  for i = 1,#gk do print(gk[i],type(os[gk[i]])) end
$$;
INFO:  time	function
INFO:  difftime	function
INFO:  execute	nil
INFO:  getenv	nil
INFO:  exit	nil
-- check that trusted can't require dangerous core modules
do language pllua $$
  print((lpcall(require,"debug")))
  print((lpcall(require,"io")))
$$;
INFO:  false
INFO:  false
--end
